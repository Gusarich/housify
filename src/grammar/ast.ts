export type AstModule = {
    kind: 'module';
    items: AstModuleItem[];
    id: number;
};

export type AstModuleItem = AstHouse;

export type AstHouse = {
    kind: 'house';
    name: AstId;
    items: AstHouseItem[];
    id: number;
};

export type AstHouseItem = AstGlobalStat | AstPlayerStat | AstHandler;

export type AstGlobalStat = {
    kind: 'globalStat';
    name: AstId;
    type: AstId;
    defaultValue?: AstExpression;
    id: number;
};

export type AstPlayerStat = {
    kind: 'playerStat';
    name: AstId;
    type: AstId;
    defaultValue?: AstExpression;
    id: number;
};

export type AstHandler = {
    kind: 'handler';
    event: AstId;
    statements: AstStatement[];
    id: number;
};

export type AstStatement =
    | AstStatementAssign
    | AstStatementLet
    | AstStatementExpression
    | AstStatementIf;

export type AstStatementAssign = {
    kind: 'statementAssign';
    lvalue: AstExpression;
    value: AstExpression;
    id: number;
};

export type AstStatementLet = {
    kind: 'statementLet';
    name: AstId;
    type: AstId;
    value: AstExpression;
    id: number;
};

export type AstStatementExpression = {
    kind: 'statementExpression';
    expression: AstExpression;
    id: number;
};

export type AstStatementIf = {
    kind: 'statementIf';
    condition: AstExpression;
    then: AstStatement[];
    else?: AstStatement[];
    id: number;
};

export type AstExpression =
    | AstExpressionBinary
    | AstExpressionUnary
    | AstIntegerLiteral
    | AstExpressionId;

export type AstBinaryOp =
    | '+'
    | '-'
    | '*'
    | '/'
    | '=='
    | '!='
    | '<'
    | '<='
    | '>'
    | '>='
    | '&&'
    | '||';

export type AstExpressionBinary = {
    kind: 'expressionBinary';
    op: AstBinaryOp;
    left: AstExpression;
    right: AstExpression;
    id: number;
};

export type AstUnaryOp = '+' | '-' | '!';

export type AstExpressionUnary = {
    kind: 'expressionUnary';
    op: AstUnaryOp;
    operand: AstExpression;
    id: number;
};

export type AstIntegerLiteral = {
    kind: 'integerLiteral';
    value: string;
    id: number;
};

export type AstExpressionId = {
    kind: 'expressionId';
    name: AstId;
    id: number;
};

export type AstId = {
    kind: 'id';
    name: string;
    id: number;
};

export type AstNode =
    | AstModule
    | AstModuleItem
    | AstHouseItem
    | AstStatement
    | AstExpression
    | AstId;

let nextId = 1;

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type DistributiveOmit<T, K extends keyof any> = T extends any
    ? Omit<T, K>
    : never;

export function createAstNode(src: DistributiveOmit<AstNode, 'id'>): AstNode {
    return Object.freeze(Object.assign({ id: nextId++ }, src));
}

export function cloneAstNode<T extends AstNode>(src: T): T {
    return { ...src, id: nextId++ };
}

export function resetNodeId() {
    nextId = 1;
}
