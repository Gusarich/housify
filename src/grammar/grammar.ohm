Housify {
    Module = ModuleItem*

    ModuleItem = House
    
    House = house id "{" HouseItem* "}"

    HouseItem = GlobalStat | PlayerStat | Handler

    GlobalStat = global id ":" Type ("=" Expression)? ";"
    PlayerStat = player id ":" Type ("=" Expression)? ";"

    Handler = handle id "{" Statement* "}"

    Statement = StatementAssign
              | StatementLet
              | StatementExpression
              | StatementIf
    
    StatementAssign = Expression "=" Expression ";"

    StatementLet = let id ":" Type "=" Expression ";"

    StatementExpression = Expression ";"

    StatementIf = if Expression "{" Statement* "}" ~else --noElse
                | if Expression "{" Statement* "}" else "{" Statement* "}" --withElse
                | if Expression "{" Statement* "}" else StatementIf --withElseIf
    
    Expression = ExpressionBinary
               | ExpressionUnary
               | ExpressionLiteral
               | ExpressionId
               | ExpressionParens
    
    ExpressionBinary = Expression binaryOp Expression
    ExpressionUnary = unaryOp Expression
    ExpressionLiteral = integerLiteral
    ExpressionId = id
    ExpressionParens = "(" Expression ")"

    binaryOp = "+" | "-" | "*" | "/" | "==" | "!=" | "<" | "<=" | ">" | ">=" | "&&" | "||"
    unaryOp = "+" | "-" | "!"

    Type = id --regular

    integerLiteral = digit+

    letterAsciiLC = "a".."z"
    letterAsciiUC = "A".."Z"
    letterAscii = letterAsciiLC | letterAsciiUC
    letterComment = letterAsciiLC | letterAsciiUC | digit | "_"

    idStart = letterAscii | "_"
    idPart = letterAscii | digit | "_"
    id = ~keyword #idStart #(idPart*)

    house = "house" ~idPart
    global = "global" ~idPart
    player = "player" ~idPart
    handle = "handle" ~idPart
    let = "let" ~idPart
    if = "if" ~idPart
    else = "else" ~idPart

    keyword = house | global | player | handle | let | if | else
}