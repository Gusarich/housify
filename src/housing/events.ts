import { ConditionType } from './conditions';
import { ActionType } from './actions';

//
// TODO: COMPLETE AND ENSURE TO MATCH ACTUAL TYPES
//

export enum EventType {
    JOIN = 'JOIN',
    QUIT = 'QUIT',
    DEATH = 'DEATH',
    KILL = 'KILL',
    RESPAWN = 'RESPAWN',
    GROUP_CHANGE = 'GROUP_CHANGE',
    PVP_STATE_CHANGE = 'PVP_STATE_CHANGE',
    FISH_CAUGHT = 'FISH_CAUGHT',
    PORTAL_USE = 'PORTAL_USE',
    DAMAGE = 'DAMAGE',
    BLOCK_BREAK = 'BLOCK_BREAK',
    PARKOUR_START = 'PARKOUR_START',
    PARKOUR_FINISH = 'PARKOUR_FINISH',
    DROP_ITEM = 'DROP_ITEM',
    PICKUP_ITEM = 'PICKUP_ITEM',
    CHANGE_HELD_ITEM = 'CHANGE_HELD_ITEM',
    TOGGLE_SNEAK = 'TOGGLE_SNEAK',
    TOGGLE_FLIGHT = 'TOGGLE_FLIGHT',
}

export type Event = {
    type: EventType;
    cancellable: boolean;
    actions: ActionType[];
    conditions: ConditionType[];
};

export const JoinEvent: Event = {
    type: EventType.JOIN,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const QuitEvent: Event = {
    type: EventType.QUIT,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const DeathEvent: Event = {
    type: EventType.DEATH,
    cancellable: true,
    actions: [],
    conditions: [],
};

export const KillEvent: Event = {
    type: EventType.KILL,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const RespawnEvent: Event = {
    type: EventType.RESPAWN,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const GroupChangeEvent: Event = {
    type: EventType.GROUP_CHANGE,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const PvpStateChangeEvent: Event = {
    type: EventType.PVP_STATE_CHANGE,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const FishCaughtEvent: Event = {
    type: EventType.FISH_CAUGHT,
    cancellable: true,
    actions: [],
    conditions: [],
};

export const PortalUseEvent: Event = {
    type: EventType.PORTAL_USE,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const DamageEvent: Event = {
    type: EventType.DAMAGE,
    cancellable: true,
    actions: [],
    conditions: [],
};

export const BlockBreakEvent: Event = {
    type: EventType.BLOCK_BREAK,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const ParkourStartEvent: Event = {
    type: EventType.PARKOUR_START,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const ParkourFinishEvent: Event = {
    type: EventType.PARKOUR_FINISH,
    cancellable: false,
    actions: [],
    conditions: [],
};

export const DropItemEvent: Event = {
    type: EventType.DROP_ITEM,
    cancellable: true,
    actions: [],
    conditions: [],
};

export const PickupItemEvent: Event = {
    type: EventType.PICKUP_ITEM,
    cancellable: true,
    actions: [],
    conditions: [],
};

export const ChangeHeldItemEvent: Event = {
    type: EventType.CHANGE_HELD_ITEM,
    cancellable: true,
    actions: [],
    conditions: [],
};

export const ToggleSneakEvent: Event = {
    type: EventType.TOGGLE_SNEAK,
    cancellable: true,
    actions: [],
    conditions: [],
};

export const ToggleFlightEvent: Event = {
    type: EventType.TOGGLE_FLIGHT,
    cancellable: true,
    actions: [],
    conditions: [],
};
