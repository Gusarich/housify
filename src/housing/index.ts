export * from './actions';
export * from './conditions';
export * from './events';
export * from './util';
