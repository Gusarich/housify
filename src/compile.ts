import { writeModule } from './generator/generate';
import { parse } from './grammar/grammar';
import { CompilerContext } from './resolver/context';
import { resolveModule } from './resolver/resolve';

export function compile(src: string, log = true) {
    // Parsing
    if (log) console.log('Parsing...');
    const moduleAst = parse(src);

    // Resolving
    if (log) console.log('Resolving...');
    const ctx = new CompilerContext();
    resolveModule(moduleAst, ctx);

    // Generating
    if (log) console.log('Generating...');
    const module = writeModule(moduleAst, ctx);

    if (log) console.log('Compilation successful!');

    return module;
}
