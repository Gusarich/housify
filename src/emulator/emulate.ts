import { EmulationError } from '../errors';
import {
    CompiledHouse,
    isGlobalStatReference,
    isStatReference,
    unwrapStatReference,
} from '../generator/generate';
import { Action, ActionKind } from '../housing/actions';
import { Condition, ConditionKind } from '../housing/conditions';
import { EventType } from '../housing/events';
import { StatChangeMode, StatCompareMode } from '../housing/util';

export class EmulatedHouse {
    handlers: Map<EventType, Action[]> = new Map();
    #globalStats: Map<string, number> = new Map();
    #playerStats: Map<string, Map<string, number>> = new Map();
    #uncollectedActions: Action[] = [];

    constructor(house: CompiledHouse) {
        for (const handler of house.handlers) {
            this.handlers.set(handler.event, handler.actions);
        }
    }

    reset() {
        this.#globalStats.clear();
        this.#playerStats.clear();
        this.#uncollectedActions = [];
    }

    emit(event: EventType, player: string) {
        if (!this.handlers.has(event)) {
            return;
        }
        const actions = this.handlers.get(event)!;
        for (const action of actions) {
            this.processAction(action, player);
        }
    }

    collect(): Action[] {
        const actions = this.#uncollectedActions;
        this.#uncollectedActions = [];
        return actions;
    }

    globalStat(stat: string): number {
        return this.#globalStats.get(stat) ?? 0;
    }

    playerStat(player: string, stat: string): number {
        return this.#playerStats.get(player)?.get(stat) ?? 0;
    }

    getValue(value: string, player: string) {
        if (isStatReference(value)) {
            const unwrapped = unwrapStatReference(value);
            if (isGlobalStatReference(value)) {
                return this.globalStat(unwrapped).toString();
            } else {
                return this.playerStat(player, unwrapped).toString();
            }
        }
        return value;
    }

    processAction(action: Action, player: string) {
        switch (action.kind) {
            case ActionKind.CHANGE_GLOBAL_STAT:
                this.changeGlobalStat(
                    player,
                    action.stat,
                    action.mode,
                    action.value,
                );
                break;
            case ActionKind.CHANGE_PLAYER_STAT:
                this.changePlayerStat(
                    player,
                    action.stat,
                    action.mode,
                    action.value,
                );
                break;
            case ActionKind.CONDITIONAL: {
                const results = action.conditions.map((condition) =>
                    this.checkCondition(condition, player),
                );
                if (
                    (action.matchAny && results.some((result) => result)) ||
                    (!action.matchAny && results.every((result) => result))
                ) {
                    for (const thenAction of action.then) {
                        this.processAction(thenAction, player);
                    }
                } else {
                    for (const elseAction of action.else) {
                        this.processAction(elseAction, player);
                    }
                }
                break;
            }
            default:
                this.#uncollectedActions.push(action);
        }
    }

    checkCondition(condition: Condition, player: string) {
        switch (condition.kind) {
            case ConditionKind.GLOBAL_STAT:
                return this.compareGlobalStat(
                    player,
                    condition.stat,
                    condition.mode,
                    condition.value,
                );
            case ConditionKind.PLAYER_STAT:
                return this.comparePlayerStat(
                    player,
                    condition.stat,
                    condition.mode,
                    condition.value,
                );
            default:
                throw new EmulationError(
                    `Unsupported condition: '${condition.kind}'`,
                );
        }
    }

    changeGlobalStat(
        player: string,
        stat: string,
        mode: StatChangeMode,
        value: string,
    ) {
        if (!this.#globalStats.has(stat)) {
            this.#globalStats.set(stat, 0);
        }

        value = this.getValue(value, player);

        switch (mode) {
            case StatChangeMode.INCREMENT:
                this.#globalStats.set(
                    stat,
                    this.#globalStats.get(stat)! + parseInt(value),
                );
                break;
            case StatChangeMode.DECREMENT:
                this.#globalStats.set(
                    stat,
                    this.#globalStats.get(stat)! - parseInt(value),
                );
                break;
            case StatChangeMode.SET:
                this.#globalStats.set(stat, parseInt(value));
                break;
            case StatChangeMode.MULTIPLY:
                this.#globalStats.set(
                    stat,
                    this.#globalStats.get(stat)! * parseInt(value),
                );
                break;
            case StatChangeMode.DIVIDE:
                this.#globalStats.set(
                    stat,
                    Math.trunc(this.#globalStats.get(stat)! / parseInt(value)),
                );
                break;
        }
    }

    changePlayerStat(
        player: string,
        stat: string,
        mode: StatChangeMode,
        value: string,
    ) {
        if (!this.#playerStats.has(player)) {
            this.#playerStats.set(player, new Map());
        }

        if (!this.#playerStats.get(player)!.has(stat)) {
            this.#playerStats.get(player)!.set(stat, 0);
        }

        value = this.getValue(value, player);

        switch (mode) {
            case StatChangeMode.INCREMENT:
                this.#playerStats
                    .get(player)!
                    .set(
                        stat,
                        this.#playerStats.get(player)!.get(stat)! +
                            parseInt(value),
                    );
                break;
            case StatChangeMode.DECREMENT:
                this.#playerStats
                    .get(player)!
                    .set(
                        stat,
                        this.#playerStats.get(player)!.get(stat)! -
                            parseInt(value),
                    );
                break;
            case StatChangeMode.SET:
                this.#playerStats.get(player)!.set(stat, parseInt(value));
                break;
            case StatChangeMode.MULTIPLY:
                this.#playerStats
                    .get(player)!
                    .set(
                        stat,
                        this.#playerStats.get(player)!.get(stat)! *
                            parseInt(value),
                    );
                break;
            case StatChangeMode.DIVIDE:
                this.#playerStats
                    .get(player)!
                    .set(
                        stat,
                        Math.trunc(
                            this.#playerStats.get(player)!.get(stat)! /
                                parseInt(value),
                        ),
                    );
                break;
        }
    }

    compareGlobalStat(
        player: string,
        stat: string,
        mode: StatCompareMode,
        value: string,
    ) {
        value = this.getValue(value, player);

        switch (mode) {
            case StatCompareMode.EQUAL:
                return (this.#globalStats.get(stat) ?? 0) === parseInt(value);
            case StatCompareMode.GREATER_THAN:
                return (this.#globalStats.get(stat) ?? 0) > parseInt(value);
            case StatCompareMode.LESS_THAN:
                return (this.#globalStats.get(stat) ?? 0) < parseInt(value);
            case StatCompareMode.GREATER_THAN_OR_EQUAL:
                return (this.#globalStats.get(stat) ?? 0) >= parseInt(value);
            case StatCompareMode.LESS_THAN_OR_EQUAL:
                return (this.#globalStats.get(stat) ?? 0) <= parseInt(value);
        }
    }

    comparePlayerStat(
        player: string,
        stat: string,
        mode: StatCompareMode,
        value: string,
    ) {
        value = this.getValue(value, player);

        switch (mode) {
            case StatCompareMode.EQUAL:
                return (
                    (this.#playerStats.get(player)?.get(stat) ?? 0) ===
                    parseInt(value)
                );
            case StatCompareMode.GREATER_THAN:
                return (
                    (this.#playerStats.get(player)?.get(stat) ?? 0) >
                    parseInt(value)
                );
            case StatCompareMode.LESS_THAN:
                return (
                    (this.#playerStats.get(player)?.get(stat) ?? 0) <
                    parseInt(value)
                );
            case StatCompareMode.GREATER_THAN_OR_EQUAL:
                return (
                    (this.#playerStats.get(player)?.get(stat) ?? 0) >=
                    parseInt(value)
                );
            case StatCompareMode.LESS_THAN_OR_EQUAL:
                return (
                    (this.#playerStats.get(player)?.get(stat) ?? 0) <=
                    parseInt(value)
                );
        }
    }
}
