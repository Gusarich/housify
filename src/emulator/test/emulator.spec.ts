import { resetNodeId } from '../../grammar/ast';
import { loadCases } from '../../utils/loadCases';
import { resetTempId } from '../../generator/generate';
import { EmulatedHouse } from '../emulate';
import { compile } from '../../compile';
import { EventType } from '../../housing/events';

describe('Emulator', () => {
    const cases = loadCases(__dirname + '/cases/');

    beforeEach(() => {
        resetNodeId();
        resetTempId();
    });

    it('should emulate counter', () => {
        const module = compile(
            cases.find((c) => c.name === 'counter')!.code,
            false,
        );
        expect(module).toMatchSnapshot();
        const house = new EmulatedHouse(module.houses[0]!);

        house.emit(EventType.JOIN, 'player1');
        house.emit(EventType.JOIN, 'player2');
        house.emit(EventType.JOIN, 'player3');
        house.emit(EventType.JOIN, 'player1');
        house.emit(EventType.JOIN, 'player2');
        house.emit(EventType.JOIN, 'player3');

        const actions = house.collect();
        expect(actions).toHaveLength(0);
        expect(house.globalStat('counter')).toBe(6);
        expect(house.playerStat('player1', 'counter')).toBe(2);
        expect(house.playerStat('player2', 'counter')).toBe(2);
        expect(house.playerStat('player3', 'counter')).toBe(2);
        expect(house.playerStat('player4', 'counter')).toBe(0);
    });

    it('should emulate arithmetics', () => {
        const module = compile(
            cases.find((c) => c.name === 'arithmetics')!.code,
            false,
        );
        expect(module).toMatchSnapshot();
        const house = new EmulatedHouse(module.houses[0]!);

        house.emit(EventType.JOIN, 'player1');
        expect(house.globalStat('counter')).toBe(23);

        house.emit(EventType.JOIN, 'player2');
        expect(house.globalStat('counter')).toBe(115);

        house.emit(EventType.JOIN, 'player3');
        expect(house.globalStat('counter')).toBe(483);

        const actions = house.collect();
        expect(actions).toHaveLength(0);
    });

    it('should emulate conditions', () => {
        const module = compile(
            cases.find((c) => c.name === 'conditions')!.code,
            false,
        );
        expect(module).toMatchSnapshot();
        const house = new EmulatedHouse(module.houses[0]!);

        house.emit(EventType.JOIN, 'player');
        expect(house.globalStat('counter')).toBe(123);

        house.emit(EventType.JOIN, 'player');
        expect(house.globalStat('counter')).toBe(456);

        house.emit(EventType.JOIN, 'player');
        expect(house.globalStat('counter')).toBe(789);

        house.emit(EventType.JOIN, 'player');
        expect(house.globalStat('counter')).toBe(0);

        house.emit(EventType.JOIN, 'player');
        expect(house.globalStat('counter')).toBe(123);

        house.emit(EventType.JOIN, 'player');
        expect(house.globalStat('counter')).toBe(456);

        house.emit(EventType.JOIN, 'player');
        expect(house.globalStat('counter')).toBe(789);

        house.emit(EventType.JOIN, 'player');
        expect(house.globalStat('counter')).toBe(0);

        const actions = house.collect();
        expect(actions).toHaveLength(0);
    });
});
