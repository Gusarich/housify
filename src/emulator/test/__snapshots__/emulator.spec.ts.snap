// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emulator should emulate arithmetics 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$0",
              "value": "3",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$1",
              "value": "2",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$2",
              "value": "%stat.global/counter%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "MULTIPLY",
              "stat": "$$2",
              "value": "%stat.player/$$1%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$3",
              "value": "%stat.player/$$0%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "INCREMENT",
              "stat": "$$3",
              "value": "%stat.player/$$2%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$4",
              "value": "1",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$5",
              "value": "%stat.player/$$3%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "DECREMENT",
              "stat": "$$5",
              "value": "%stat.player/$$4%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$6",
              "value": "4",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$7",
              "value": "%stat.player/$$5%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "INCREMENT",
              "stat": "$$7",
              "value": "%stat.player/$$6%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$8",
              "value": "5",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$9",
              "value": "%stat.player/$$7%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "INCREMENT",
              "stat": "$$9",
              "value": "%stat.player/$$8%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$x",
              "value": "%stat.player/$$9%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$0",
              "value": "1",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$1",
              "value": "2",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$2",
              "value": "%stat.player/$x%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "MULTIPLY",
              "stat": "$$2",
              "value": "%stat.player/$$1%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$3",
              "value": "%stat.player/$$0%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "INCREMENT",
              "stat": "$$3",
              "value": "%stat.player/$$2%",
            },
            {
              "kind": "CHANGE_GLOBAL_STAT",
              "mode": "SET",
              "stat": "counter",
              "value": "%stat.player/$$3%",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "TestHouse",
      "playerStats": [],
    },
  ],
  "kind": "module",
}
`;

exports[`Emulator should emulate conditions 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$0",
              "value": "0",
            },
            {
              "conditions": [
                {
                  "kind": "GLOBAL_STAT",
                  "mode": "EQUAL",
                  "stat": "counter",
                  "value": "%stat.player/$$0%",
                },
              ],
              "else": [
                {
                  "kind": "CHANGE_PLAYER_STAT",
                  "mode": "SET",
                  "stat": "$$1",
                  "value": "0",
                },
              ],
              "kind": "CONDITIONAL",
              "matchAny": false,
              "then": [
                {
                  "kind": "CHANGE_PLAYER_STAT",
                  "mode": "SET",
                  "stat": "$$1",
                  "value": "1",
                },
              ],
            },
            {
              "conditions": [
                {
                  "kind": "PLAYER_STAT",
                  "mode": "EQUAL",
                  "stat": "$$1",
                  "value": "1",
                },
              ],
              "else": [
                {
                  "kind": "CHANGE_PLAYER_STAT",
                  "mode": "SET",
                  "stat": "$$0",
                  "value": "123",
                },
                {
                  "conditions": [
                    {
                      "kind": "GLOBAL_STAT",
                      "mode": "EQUAL",
                      "stat": "counter",
                      "value": "%stat.player/$$0%",
                    },
                  ],
                  "else": [
                    {
                      "kind": "CHANGE_PLAYER_STAT",
                      "mode": "SET",
                      "stat": "$$1",
                      "value": "0",
                    },
                  ],
                  "kind": "CONDITIONAL",
                  "matchAny": false,
                  "then": [
                    {
                      "kind": "CHANGE_PLAYER_STAT",
                      "mode": "SET",
                      "stat": "$$1",
                      "value": "1",
                    },
                  ],
                },
                {
                  "conditions": [
                    {
                      "kind": "PLAYER_STAT",
                      "mode": "EQUAL",
                      "stat": "$$1",
                      "value": "1",
                    },
                  ],
                  "else": [
                    {
                      "kind": "CHANGE_PLAYER_STAT",
                      "mode": "SET",
                      "stat": "$$0",
                      "value": "456",
                    },
                    {
                      "conditions": [
                        {
                          "kind": "GLOBAL_STAT",
                          "mode": "EQUAL",
                          "stat": "counter",
                          "value": "%stat.player/$$0%",
                        },
                      ],
                      "else": [
                        {
                          "kind": "CHANGE_PLAYER_STAT",
                          "mode": "SET",
                          "stat": "$$1",
                          "value": "0",
                        },
                      ],
                      "kind": "CONDITIONAL",
                      "matchAny": false,
                      "then": [
                        {
                          "kind": "CHANGE_PLAYER_STAT",
                          "mode": "SET",
                          "stat": "$$1",
                          "value": "1",
                        },
                      ],
                    },
                    {
                      "conditions": [
                        {
                          "kind": "PLAYER_STAT",
                          "mode": "EQUAL",
                          "stat": "$$1",
                          "value": "1",
                        },
                      ],
                      "else": [
                        {
                          "kind": "CHANGE_PLAYER_STAT",
                          "mode": "SET",
                          "stat": "$$0",
                          "value": "0",
                        },
                        {
                          "kind": "CHANGE_GLOBAL_STAT",
                          "mode": "SET",
                          "stat": "counter",
                          "value": "%stat.player/$$0%",
                        },
                      ],
                      "kind": "CONDITIONAL",
                      "matchAny": false,
                      "then": [
                        {
                          "kind": "CHANGE_PLAYER_STAT",
                          "mode": "SET",
                          "stat": "$$0",
                          "value": "789",
                        },
                        {
                          "kind": "CHANGE_GLOBAL_STAT",
                          "mode": "SET",
                          "stat": "counter",
                          "value": "%stat.player/$$0%",
                        },
                      ],
                    },
                  ],
                  "kind": "CONDITIONAL",
                  "matchAny": false,
                  "then": [
                    {
                      "kind": "CHANGE_PLAYER_STAT",
                      "mode": "SET",
                      "stat": "$$0",
                      "value": "456",
                    },
                    {
                      "kind": "CHANGE_GLOBAL_STAT",
                      "mode": "SET",
                      "stat": "counter",
                      "value": "%stat.player/$$0%",
                    },
                  ],
                },
              ],
              "kind": "CONDITIONAL",
              "matchAny": false,
              "then": [
                {
                  "kind": "CHANGE_PLAYER_STAT",
                  "mode": "SET",
                  "stat": "$$0",
                  "value": "123",
                },
                {
                  "kind": "CHANGE_GLOBAL_STAT",
                  "mode": "SET",
                  "stat": "counter",
                  "value": "%stat.player/$$0%",
                },
              ],
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "MyHouse",
      "playerStats": [],
    },
  ],
  "kind": "module",
}
`;

exports[`Emulator should emulate counter 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$0",
              "value": "1",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$1",
              "value": "%stat.global/counter%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "INCREMENT",
              "stat": "$$1",
              "value": "%stat.player/$$0%",
            },
            {
              "kind": "CHANGE_GLOBAL_STAT",
              "mode": "SET",
              "stat": "counter",
              "value": "%stat.player/$$1%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$0",
              "value": "1",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$1",
              "value": "%stat.player/counter%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "INCREMENT",
              "stat": "$$1",
              "value": "%stat.player/$$0%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "counter",
              "value": "%stat.player/$$1%",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "Counter",
      "playerStats": [
        "counter",
      ],
    },
  ],
  "kind": "module",
}
`;

exports[`Emulator should emulate multiple handlers 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$0",
              "value": "1",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$1",
              "value": "%stat.global/counter%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "INCREMENT",
              "stat": "$$1",
              "value": "%stat.player/$$0%",
            },
            {
              "kind": "CHANGE_GLOBAL_STAT",
              "mode": "SET",
              "stat": "counter",
              "value": "%stat.player/$$1%",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
        {
          "actions": [
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$0",
              "value": "1",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "SET",
              "stat": "$$1",
              "value": "%stat.global/counter%",
            },
            {
              "kind": "CHANGE_PLAYER_STAT",
              "mode": "DECREMENT",
              "stat": "$$1",
              "value": "%stat.player/$$0%",
            },
            {
              "kind": "CHANGE_GLOBAL_STAT",
              "mode": "SET",
              "stat": "counter",
              "value": "%stat.player/$$1%",
            },
          ],
          "event": "QUIT",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "MyHouse",
      "playerStats": [],
    },
  ],
  "kind": "module",
}
`;
